### PR TITLE
KAFKA-15544: Enable integration tests for new consumer

### PR DESCRIPTION
**Pull request by AndrewJSchofield**
This PR parameterizes the consumer integration tests so they can be run against
the existing "generic" group protocol and the new "consumer" group protocol
introduced in KIP-848.

The KIP-848 client code is under construction so some of the tests do not run on
both variants to start with, but the idea is that the tests can be enabled as the gaps
in functionality are closed.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (kirktrue) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: So this is incorrect in `trunk`, right?

File: core/src/test/scala/integration/kafka/api/AbstractConsumerTest.scala
Comment (kirktrue) [inline core/src/test/scala/integration/kafka/api/AbstractConsumerTest.scala]: At some point we need to test that if the broker's configuration has this set to `false` that the client won't up and explode. That's a future task...

File: core/src/test/scala/integration/kafka/api/BaseAsyncConsumerTest.scala
Comment (kirktrue) [inline core/src/test/scala/integration/kafka/api/BaseAsyncConsumerTest.scala]: Good riddance!

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala
Comment (kirktrue) [inline core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: Is it possible to have a different type of 'source' that can be defined once vs. on each test? More of a nit, but curious.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (kirktrue) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:1900]: This ain't Java, come on!

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (kirktrue) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Why don't we need this now?

**Review by kirktrue — APPROVED** [review]
Thanks for the PR @AndrewJSchofield!

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (AndrewJSchofield) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: In this context, `servers` and `brokers` are interchangeable. This change makes the test work for ZK or KRaft. Previously, it was ZK-only.

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala
Comment (AndrewJSchofield) [reply core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: Perhaps, but I want to be able to have a different array for each test to enable them to be turned on individually.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (AndrewJSchofield) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: I'm not convinced we ever did. Again, this is ZK-specific. The tests work on both variants without this check.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (AndrewJSchofield) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:1916]: I missed one :)

**Comment by kirktrue** [comment]
@philipnee can you tag with `ctr` and `KIP-848` 🥺

**Comment by AndrewJSchofield** [comment]
Converting back to draft. This PR depends on https://github.com/apache/kafka/pull/14801 being merged. This is ready for review, but will not build cleanly until the other PR lands.

**Comment by kirktrue** [comment]
@cadonna / @lucasbru—can we add the `ctr` label to this PR?

**Comment by cadonna** [comment]
Done!

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: This is just a transient state, isn't it? At least https://issues.apache.org/jira/browse/KAFKA-15327 says that committing on close is planned. If KAFKA-15327 is still valid, can we formulate this comment accordingly like `close() is not committing offsets in consumer group protocol for now but it should when implementation is complete`.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: See my comment above about the functionality being supported in future.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: I guess the `ConsumerRebalanceListener` is also something that will be supported in the future by the consumer group protocol.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Same here

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Will also interceptors be supported?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: See my other comments.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: See my other comments.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: See my other comments

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: See my other comments.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: See my other comments

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: See my other comments

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: I agree with @AndrewJSchofield . Method `createTopics()` already verifies that the topic was created.

File: core/src/test/scala/kafka/utils/TestInfoUtils.scala
Comment (cadonna) [inline core/src/test/scala/kafka/utils/TestInfoUtils.scala]: Could explain why this is needed?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Static membership will be supported in future, right?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:2083]: Could you make this clear in the name of the test instead of using a inline comment?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:2102]: See my comment above

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: See my other comments about functionality that will be supported in the future.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: This name suggests that a group ID of `null` is not supported in general with manual assignment. However, `null` is supported as group ID with manual assignment. Just commits are not supported with group ID = `null`. Could you please use a more descriptive name for this test like `testNullGroupIdNotSupportedIfCommitting`?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (cadonna) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:2268]: Wouldn't it be better to consumer and commit some records before retrieving the committed offsets a second time?

**Review by cadonna — COMMENTED** [review]
Thanks for the PR, @AndrewJSchofield !

Here my comments.

**Comment by dajac** [comment]
@AndrewJSchofield We are about to merge https://github.com/apache/kafka/pull/14781. Is it possible to build this one on top of it?

**Comment by AndrewJSchofield** [comment]
@dajac I suggest you simply merge #14871 and let me deal with the fall-out. This one is not going to be far behind.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (AndrewJSchofield) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: I agree with the idea, but I will use a less wordy formulation such as "temporarily". I expect a bunch of these will be resolved in the next two weeks and we can run most of the tests across both consumers.

File: core/src/test/scala/kafka/utils/TestInfoUtils.scala
Comment (AndrewJSchofield) [reply core/src/test/scala/kafka/utils/TestInfoUtils.scala]: It's because the parameterized tests in `PlaintextConsumerTest` choose between "generic" and "consumer". The new consumer ("consumer") is only supported for KRaft clusters, so it's necessary to make the test use KRaft in this instance.

**Comment by dajac** [comment]
I just merged https://github.com/apache/kafka/pull/14781. Thanks!

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala
Comment (AndrewJSchofield) [reply core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: This has been replaced with a `MethodSource` that is capable of returning whatever combination we want.

File: core/src/test/scala/kafka/utils/TestInfoUtils.scala
Comment (AndrewJSchofield) [reply core/src/test/scala/kafka/utils/TestInfoUtils.scala]: I've replaced this in the new commit.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (AndrewJSchofield) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala:2268]: I didn't write this test and there's no comment by the author. I suppose this was done specifically to catch a problem and I prefer not to mess around with it when I don't understand.

**Comment by AndrewJSchofield** [comment]
This PR now reflects the changes in KAFKA-14781 and also tests the new consumer.

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala
Comment (dajac) [inline core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:125]: We also need to test the `generic` with `kraft+kip848`. This is what all the tests with `kraft+kip848` prior to your change did.

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala
Comment (dajac) [inline core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: We likely need it here too.

Comment (AndrewJSchofield) [reply core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: OK. Nice and easy to change now I've refactored it. I'll get on it :)

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala
Comment (dajac) [inline core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: Adding this change to all the tests is a bit annoying. Have you considered adding this to `IntegrationTestHarness.doSetup` or in `createConsumer`? We could infer it like we did with `isNewGroupCoordinatorEnabled()` in the same class.

Comment (AndrewJSchofield) [reply core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: I did consider that. I have already endured the annoyance. I'll take a look.

Comment (AndrewJSchofield) [reply core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: Done

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala
Comment (lucasbru) [inline core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: Can we remove this line now?

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala
Comment (lucasbru) [reply core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: Can we remove this line now?

File: core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala
Comment (lucasbru) [inline core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala]: can we remove this line now?

**Review by lucasbru — COMMENTED** [review]
Scala 2.12 build needs to be fixed, and I left some minor comments. But generally this looks good to me

**Review by lucasbru — APPROVED** [review]

**Comment by lucasbru** [comment]
CI is blocked by a hanging test bug that is going to celebrate its fourth birthday soon: https://issues.apache.org/jira/browse/KAFKA-9470. I restarted CI and submitted https://github.com/apache/kafka/pull/14855

**Comment by lucasbru** [comment]
I created 3 new flaky test tickets, but all failed tests were flaky before on master.

**Pull request merged**
